### PR TITLE
Spring Security added

### PR DESCRIPTION
**Pull request by Rommelua**

### CODE REVIEW

**Review by aleksandr-hrankin — APPROVED** [review]

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@EnableWebSecurity
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+    @Autowired
+    public void configureGlobal(AuthenticationManagerBuilder builder) throws Exception {
+        builder.inMemoryAuthentication()
+                .withUser("user")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java:17]: does it work ok without `.passwordEncoder()`?

File: src/main/java/com/dev/cinema/config/SecurityInitializer.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.dev.cinema.config;
+
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.web.context.AbstractSecurityWebApplicationInitializer;
+
+@EnableWebSecurity
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/config/SecurityInitializer.java]:
```suggestion
```

File: src/main/java/com/dev/cinema/model/dto/UserRequestDto.java [hunk 3]
```diff
@@ -1,8 +1,24 @@
...
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+
+@FieldsValueMatch.List({
+        @FieldsValueMatch(
+                field = "password",
+                fieldMatch = "passwordRepeat",
+                message = "Passwords do not match!"
+        )
+})
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/UserRequestDto.java]:
```suggestion
@FieldsValueMatch(
                field = "password",
                fieldMatch = "passwordRepeat",
                message = "Passwords do not match!"
)
```

what if you try this and try removing `@List` from inside `@FieldsValueMatch`?

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -35,4 +37,10 @@ public User get(Long id) {
     public Optional<User> findByEmail(String email) {
         return userDao.findByEmail(email);
     }
+
+    @Override
+    public User getUser(Authentication authentication) {
+        UserDetails details = (UserDetails) authentication.getPrincipal();
+        return findByEmail(details.getUsername()).get();
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: this is not part of the application business logic and should be handled on the controller layer

also, have you tried taking `authentication.getName()`?

File: src/main/java/com/dev/cinema/validation/EmailValidator.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<Email, String> {
+    private static final String REGEXP
+            = "(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x0"
+              + "8\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e"
+              + "-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-"
+              + "9])?|\\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\\.){3}(?:(2(5[0-5]|["
+              + "0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\"
+              + "x0e-\\x1f\\x21-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)])";
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/validation/EmailValidator.java:13]: looks too complex for the average usage, how do you think?

File: src/main/java/com/dev/cinema/validation/EmailValidator.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+    @Override
+    public boolean isValid(String email, ConstraintValidatorContext constraintValidatorContext) {
+        return email.matches(REGEXP);
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/validation/EmailValidator.java]: let's add null check

File: src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        Object fieldValue = new BeanWrapperImpl(value)
+                .getPropertyValue(field);
+        Object fieldMatchValue = new BeanWrapperImpl(value)
+                .getPropertyValue(fieldMatch);
+        if (fieldValue != null) {
+            return fieldValue.equals(fieldMatchValue);
+        } else {
+            return fieldMatchValue == null;
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java]: you can use ternary operator here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
I would suggest to add validation of other requestDtos as well

File: src/main/java/com/dev/cinema/config/SecurityConfig.java (hunk 1 above)
Comment (Rommelua) [reply src/main/java/com/dev/cinema/config/SecurityConfig.java:17]: Yes, it does

File: src/main/java/com/dev/cinema/model/dto/UserRequestDto.java (hunk 3 above)
Comment (Rommelua) [reply src/main/java/com/dev/cinema/model/dto/UserRequestDto.java]: done

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java (hunk 4 above)
Comment (Rommelua) [reply src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java]: done

File: src/main/java/com/dev/cinema/validation/EmailValidator.java (hunk 5 above)
Comment (Rommelua) [reply src/main/java/com/dev/cinema/validation/EmailValidator.java:13]: No, I dont think so, I saw 4 times longer!
javax.validation has its @Email annotation, but if we need create our own...
This regex has **2542** likes at https://stackoverflow.com/questions/201323/how-to-validate-an-email-address-using-a-regular-expression

File: src/main/java/com/dev/cinema/validation/EmailValidator.java (hunk 6 above)
Comment (Rommelua) [reply src/main/java/com/dev/cinema/validation/EmailValidator.java]: done

File: src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java (hunk 7 above)
Comment (Rommelua) [reply src/main/java/com/dev/cinema/validation/FieldsValueMatchValidator.java]: I've done even better - Objects.equals

File: src/main/java/com/dev/cinema/model/dto/MovieSessionRequestDto.java [hunk 8]
```diff
@@ -2,10 +2,14 @@
...
 import com.fasterxml.jackson.annotation.JsonFormat;
 import java.time.LocalDateTime;
+import javax.validation.constraints.NotNull;
 
 public class MovieSessionRequestDto {
+    @NotNull(message = "MovieId can't be null")
     private Long movieId;
+    @NotNull(message = "CinemaHallId can't be null")
     private Long cinemaHallId;
+    @NotNull(message = "ShowTime can't be null")
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/MovieSessionRequestDto.java:12]: does it work without @Valid?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**